### PR TITLE
Offline (FBS) queue system

### PR DESCRIPTION
**Pull request by cableman**

### CODE REVIEW

File: engine/plugins/queue/queue.js [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+    bus.on('queue.add.checkout', function(obj) {
+        // Ensure that it's an copy of the object.
+        const data = JSON.parse(JSON.stringify(obj));
+        data.queued = true;
+
+        queue.add('checkout', data);
+    });
+
+    bus.on('queue.add.checkin', function(obj) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:300]: Missing comment

File: engine/plugins/queue/queue.js [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+ * @param {array} imports
+ *   The other plugins available.
+ * @param {function} register
+ *   Callback function used to register this plugin.
+ */
+module.exports = function(options, imports, register) {
+    const bus = imports.bus;
+    const queue = new Queue(bus, options.host, options.port, options.db);
+
+    bus.on('queue.add.checkout', function(obj) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:292]: Missing comment

File: engine/plugins/queue/queue.js [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                    message: require('util').inspect(res, true, 10)
+                });
+                done(new Error(res.screenMessage));
+            }
+
+            // Success the item have been checked-out.
+            done(null, res);
+        });
+
+        this.bus.once(data.errorEvent, function(err) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:260]: Missing comment

File: engine/plugins/queue/queue.js [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+     */
+    checkout(job, done) {
+        const data = job.data;
+        const self = this;
+
+        // Set event callbacks that FBS should use.
+        data.busEvent = 'queue.fbs.checkout.success' + data.itemIdentifier;
+        data.errorEvent = 'queue.fbs.checkout.error' + data.itemIdentifier;
+
+        this.bus.once(data.busEvent, function(res) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:247]: Missing comment

File: engine/plugins/queue/queue.js [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                    message: require('util').inspect(res, true, 10)
+                });
+                done(new Error(res.screenMessage));
+            }
+
+            // Success the item have been checked-in.
+            done(null, res);
+        });
+
+        this.bus.once(data.errorEvent, function(err) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:211]: Missing comment

File: engine/plugins/queue/queue.js [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+     */
+    checkin(job, done) {
+        const data = job.data;
+        const self = this;
+
+        // Set event callbacks that FBS should use.
+        data.busEvent = 'queue.fbs.checkout.success' + data.itemIdentifier;
+        data.errorEvent = 'queue.fbs.checkout.error' + data.itemIdentifier;
+
+        this.bus.once(data.busEvent, function(res) {
```
Comment (tuj) [inline engine/plugins/queue/queue.js:198]: Missing comment

**Review by tuj — CHANGES_REQUESTED** [review]

File: engine/plugins/queue/queue.js [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+     *
+     * @param type
+     *  The type of queue (checkin or checkout).
+     *
+     * @returns {*}
+     *   The queue if found. If not null.
+     *
+     * @private
+     */
+    _findQueue(type) {
```
[the remark this answers is not in the document]
Comment (cableman) [reply engine/plugins/queue/queue.js:97]: Private function.

File: engine/plugins/server/server.js [hunk 8]
```diff
@@ -92,6 +92,16 @@ module.exports = function(options, imports, register) {
...
         });
 
+        /**
+         * Handle FBS is offline events.
+         *
+         * @TODO: Check that queue error is handled. It means that the
+         *        queue/offline system is not working.
+         */
+        bus.on('queue.error', () => {
+            socket.emit('Offline');
```
[the remark this answers is not in the document]
Comment (cableman) [reply engine/plugins/server/server.js:102]: Changed to `OutOfOrder`

File: engine/plugins/queue/queue.js [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+ *   The other plugins available.
+ * @param {function} register
+ *   Callback function used to register this plugin.
+ */
+module.exports = function(options, imports, register) {
+    const bus = imports.bus;
+    const queue = new Queue(bus, options.host, options.port, options.db);
+
+    bus.on('queue.add.checkout', function(obj) {
+        // Ensure that it's an copy of the object.
```
[the remark this answers is not in the document]
Comment (cableman) [reply engine/plugins/queue/queue.js:293]: Fixed

File: engine/plugins/queue/queue.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (cableman) [reply engine/plugins/queue/queue.js:97]: Other languages often permit private properties to be declared. That’s not possible in ES6, so developers often work around it using the underscore convention (_propertyName),

[the remark this answers is not in the document]
Comment (cableman) [reply engine/plugins/queue/queue.js:97]: Do this get PHP Storm to call it private in its autocomplete.

File: engine/plugins/queue/queue.js (hunk 1 above)
Comment (cableman) [reply engine/plugins/queue/queue.js:300]: Fixed

File: engine/plugins/queue/queue.js (hunk 2 above)
Comment (cableman) [reply engine/plugins/queue/queue.js:292]: Fixed

File: engine/plugins/queue/queue.js (hunk 3 above)
Comment (cableman) [reply engine/plugins/queue/queue.js:260]: Don't think it is need in these calls... it is not something to do with the plugins "API" as with the specific events in the registry function.

This only an async function call... an the comment would be the same for them all..

File: engine/plugins/queue/queue.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (tuj) [reply engine/plugins/queue/queue.js:97]: ok

File: engine/plugins/queue/queue.js (hunk 3 above)
Comment (tuj) [reply engine/plugins/queue/queue.js:260]: // Handle error event.

**Review by tuj — APPROVED** [review]

**Pull request merged**